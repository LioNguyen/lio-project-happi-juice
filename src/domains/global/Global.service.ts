import { IGlobalActions, IShowToastParams } from "./Global.types";

class GlobalService {
  private static handlers: Partial<IGlobalActions> = {};

  static setHandlers(actions: Partial<IGlobalActions>) {
    this.handlers = actions;
  }

  static showToast(params: IShowToastParams) {
    this.handlers.showToast?.(params);
  }

  static showLoader() {
    this.handlers.showLoader?.();
  }

  static hideLoader() {
    this.handlers.hideLoader?.();
  }
}

export { GlobalService };
