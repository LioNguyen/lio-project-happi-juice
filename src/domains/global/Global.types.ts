const TOAST_TYPE = {
  success: 'success',
  error: 'error',
  info: 'info',
  warning: 'warning',
} as const

type TToastTypeValue = (typeof TOAST_TYPE)[keyof typeof TOAST_TYPE]

interface IToastMessage {
  id: string
  type: TToastTypeValue
  message: string
  showIcon?: boolean
}

interface IShowToastParams {
  type: TToastTypeValue
  message: string
  showIcon?: boolean
}

// START: Modal config
const MODAL_NAME = {
  orderConfirm: 'orderConfirm',
} as const

interface IOpenModalParams extends IModalConfig {
  name: (typeof MODAL_NAME)[keyof typeof MODAL_NAME]
}

interface IModalConfig {
  data?: any
  size?: 'sm' | 'md' | 'lg' | 'xl' | '2xl' | 'full'
}

interface IModalInstance extends IModalConfig {
  isOpen: boolean
}

type TModalMap = Record<string, IModalInstance>
// END: Modal config

// START: Sheet config
const SHEET_NAME = {
  orderFormMobile: 'orderFormMobile',
} as const

interface IOpenSheetParams extends ISheetConfig {
  name: (typeof SHEET_NAME)[keyof typeof SHEET_NAME]
}

interface ISheetConfig {
  data?: any
  side?: 'left' | 'right'
}

interface ISheetInstance extends ISheetConfig {
  isOpen: boolean
  order: number // to track sheets order
}

type TSheetMap = Record<string, ISheetInstance>
// END: Sheet config

interface IGlobalStore {
  toasts: IToastMessage[]
  isLoading: boolean
  modals: {
    instances: TModalMap
    activeModals: string[] // array of modal names
  }
  sheets: {
    instances: TSheetMap
    activeSheets: string[] // array of sheet names in order
  }
}

interface IGlobalActions {
  showToast: (params: IShowToastParams) => void
  closeToast: (id: string) => void
  showLoader: () => void
  hideLoader: () => void
  // Modal actions
  openModal: (params: IOpenModalParams) => void
  closeModal: (name: string) => void
  closeAllModals: () => void
  // Sheet actions
  openSheet: (params: IOpenSheetParams) => void
  closeSheet: (name: string) => void
  closeAllSheets: () => void
}

type TGlobalContextValue = IGlobalStore & IGlobalActions

export { MODAL_NAME, SHEET_NAME, TOAST_TYPE }

export type {
  IGlobalActions,
  IGlobalStore,
  // Modal
  IModalConfig,
  IModalInstance,
  IOpenModalParams,

  // Sheet
  ISheetConfig,
  ISheetInstance,
  IOpenSheetParams,

  // Others
  IShowToastParams,
  IToastMessage,
  TGlobalContextValue,
  TModalMap,
  TSheetMap,
  TToastTypeValue,
}
