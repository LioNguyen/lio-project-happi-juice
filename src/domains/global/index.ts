export * from "./Global.context";
export * from "./Global.service";
export * from "./Global.types";
