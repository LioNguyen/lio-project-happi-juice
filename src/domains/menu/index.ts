export * from './Menu.constants'
export * from './Menu.store'
export * from './Menu.types'
