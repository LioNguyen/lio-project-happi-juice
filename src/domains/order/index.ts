export * from './Order.hooks'
export * from './Order.service'
