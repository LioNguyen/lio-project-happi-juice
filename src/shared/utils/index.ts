export * from './lazyLoad'
export * from './tailwind'
