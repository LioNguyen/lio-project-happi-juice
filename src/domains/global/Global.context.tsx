import {
  useCallback,
  useEffect,
  useMemo,
  useState,
  type ReactNode,
} from 'react'

import { ModalContainer } from '@/components/appModal'
import { SheetContainer } from '@/components/appSheet'
import { createContext } from '@/shared/utils'
import { Loader } from '@designSystem/components/loader'
import { Toast } from '@designSystem/components/toast'
import { GlobalService } from './Global.service'
import {
  IGlobalStore,
  IOpenModalParams,
  IOpenSheetParams,
  IShowToastParams,
  IToastMessage,
  TGlobalContextValue,
} from './Global.types'

const initialStore: IGlobalStore = {
  toasts: [],
  isLoading: false,
  modals: {
    instances: {},
    activeModals: [],
  },
  sheets: {
    instances: {},
    activeSheets: [],
  },
}

// Create context with custom utility
const [Provider, useGlobal] = createContext<TGlobalContextValue>()

const GlobalProvider = ({ children }: { children: ReactNode }) => {
  const [store, setStore] = useState<IGlobalStore>(initialStore)

  // START: Toast handler
  const showToast = useCallback(
    ({ type, message, showIcon = true }: IShowToastParams) => {
      const id = Date.now().toString()
      const newToast: IToastMessage = {
        id,
        type,
        message,
        showIcon,
      }

      setStore((prev) => ({
        ...prev,
        toasts: [...prev.toasts, newToast],
      }))

      setTimeout(() => {
        closeToast(id)
      }, 3000)
      // eslint-disable-next-line react-hooks/exhaustive-deps
    },
    [],
  )

  const closeToast = useCallback((id: string) => {
    setStore((prev) => ({
      ...prev,
      toasts: prev.toasts.filter((toast) => toast.id !== id),
    }))
  }, [])
  // END: Toast handler

  // START: Loader handler
  const showLoader = useCallback(() => {
    setStore((prev) => ({ ...prev, isLoading: true }))
  }, [])

  const hideLoader = useCallback(() => {
    setStore((prev) => ({ ...prev, isLoading: false }))
  }, [])
  // END: Loader handler

  // START: Modal handlers
  const openModal = useCallback(({ name, ...config }: IOpenModalParams) => {
    setStore((prev) => {
      const activeModals = [...prev.modals.activeModals]

      if (!activeModals.includes(name)) {
        activeModals.push(name)
      }

      return {
        ...prev,
        modals: {
          instances: {
            ...prev.modals.instances,
            [name]: {
              ...config,
              isOpen: true,
            },
          },
          activeModals,
        },
      }
    })
  }, [])

  const closeModal = useCallback((name: string) => {
    setStore((prev) => {
      const activeModals = prev.modals.activeModals.filter(
        (modal) => modal !== name,
      )
      const instances = { ...prev.modals.instances }

      if (instances[name]) {
        instances[name] = {
          ...instances[name],
          isOpen: false,
        }
      }

      return {
        ...prev,
        modals: {
          instances,
          activeModals,
        },
      }
    })

    // Clean up after animation
    setTimeout(() => {
      setStore((prev) => {
        const instances = { ...prev.modals.instances }
        delete instances[name]

        return {
          ...prev,
          modals: {
            ...prev.modals,
            instances,
          },
        }
      })
    }, 200)
  }, [])

  const closeAllModals = useCallback(() => {
    setStore((prev) => ({
      ...prev,
      modals: {
        instances: Object.keys(prev.modals.instances).reduce(
          (acc, key) => ({
            ...acc,
            [key]: {
              ...prev.modals.instances[key],
              isOpen: false,
            },
          }),
          {},
        ),
        activeModals: [],
      },
    }))
  }, [])
  // END: Modal handlers

  // START: Sheet handlers
  const openSheet = useCallback(({ name, ...config }: IOpenSheetParams) => {
    setStore((prev) => {
      const activeSheets = [...prev.sheets.activeSheets]
      const sheetExists = name in prev.sheets.instances

      if (!activeSheets.includes(name)) {
        activeSheets.push(name)
      }

      // Update all existing sheets to increase their order by 1
      const updatedInstances = Object.entries(prev.sheets.instances).reduce(
        (acc, [key, sheet]) => ({
          ...acc,
          [key]: {
            ...sheet,
            order: sheetExists ? sheet.order : sheet.order + 1,
          },
        }),
        {},
      )

      return {
        ...prev,
        sheets: {
          instances: {
            ...updatedInstances,
            [name]: {
              ...config,
              isOpen: true,
              order: 0, // New sheet always gets order 0
            },
          },
          activeSheets,
        },
      }
    })
  }, [])

  const closeSheet = useCallback((name: string) => {
    setStore((prev) => {
      const closedSheetOrder = prev.sheets.instances[name]?.order || 0
      const activeSheets = prev.sheets.activeSheets.filter(
        (sheet) => sheet !== name,
      )
      const instances = { ...prev.sheets.instances }

      // Decrease order for sheets that were after the closed sheet
      Object.keys(instances).forEach((sheetName) => {
        if (instances[sheetName].order > closedSheetOrder) {
          instances[sheetName] = {
            ...instances[sheetName],
            order: instances[sheetName].order - 1,
          }
        }
      })

      if (instances[name]) {
        instances[name] = {
          ...instances[name],
          isOpen: false,
        }
      }

      return {
        ...prev,
        sheets: {
          instances,
          activeSheets,
        },
      }
    })

    setTimeout(() => {
      setStore((prev) => {
        const activeSheets = prev.sheets.activeSheets.filter(
          (sheet) => sheet !== name,
        )
        const instances = { ...prev.sheets.instances }

        if (instances[name]) {
          delete instances[name]
        }

        return {
          ...prev,
          sheets: {
            instances,
            activeSheets,
          },
        }
      })
    }, 200)
  }, [])

  const closeAllSheets = useCallback(() => {
    setStore((prev) => ({
      ...prev,
      sheets: {
        instances: Object.keys(prev.sheets.instances).reduce(
          (acc, key) => ({
            ...acc,
            [key]: {
              ...prev.sheets.instances[key],
              isOpen: false,
            },
          }),
          {},
        ),
        activeSheets: [],
      },
    }))
  }, [])
  // END: Sheet handlers

  // Register handlers with service
  useEffect(() => {
    const handlers = { showToast, closeToast, showLoader, hideLoader }

    GlobalService.setHandlers(handlers)
  }, [showToast, closeToast, showLoader, hideLoader])

  const contextValue: TGlobalContextValue = useMemo(
    () => ({
      ...store,
      showToast,
      closeToast,
      showLoader,
      hideLoader,

      // Modal
      openModal,
      closeModal,
      closeAllModals,

      // Sheet
      openSheet,
      closeSheet,
      closeAllSheets,
    }),
    [
      store,
      showToast,
      closeToast,
      showLoader,
      hideLoader,
      openModal,
      closeModal,
      closeAllModals,
      openSheet,
      closeSheet,
      closeAllSheets,
    ],
  )

  return (
    <Provider value={contextValue}>
      <Loader isLoading={store.isLoading} />
      <Toast toasts={store.toasts} closeToast={closeToast} />
      <ModalContainer />
      <SheetContainer />

      {children}
    </Provider>
  )
}

export { GlobalProvider, useGlobal }
