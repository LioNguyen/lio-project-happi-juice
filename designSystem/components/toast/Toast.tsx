import { AlertTriangle, CheckCircle, Info, XCircle } from "lucide-react";
import { useState } from "react";

import { IToastMessage, TOAST_TYPE, TToastTypeValue } from "@/domains/global";

interface IToastProps {
  toasts: IToastMessage[];
  closeToast: (id: string) => void;
}

const TOAST_DURATION = 200; // milliseconds

const Toast = ({ toasts, closeToast }: IToastProps) => {
  const [exitingToasts, setExitingToasts] = useState<string[]>([]);

  // Handle close toast with animation
  const handleClose = (id: string) => {
    // Add toast into trigger exit animation list
    setExitingToasts(prev => [...prev, id]);

    // Wait animation complete before remove toast
    setTimeout(() => {
      closeToast(id);
      // Remove toast from list
      setExitingToasts(prev => prev.filter(toastId => toastId !== id));
    }, TOAST_DURATION);
  };

  // Icon Component for toast
  const ToastIcon = ({ toast }: { toast: IToastMessage }) => {
    if (toast.showIcon === false) return null;

    const iconProps = {
      className: "shrink-0",
      size: 20,
    };

    switch (toast.type) {
      case TOAST_TYPE.success:
        return <CheckCircle data-testid="CheckCircle-icon" {...iconProps} />;
      case TOAST_TYPE.error:
        return <XCircle data-testid="XCircle-icon" {...iconProps} />;
      case TOAST_TYPE.warning:
        return <AlertTriangle data-testid="AlertTriangle-icon" {...iconProps} />;
      case TOAST_TYPE.info:
        return <Info data-testid="Info-icon" {...iconProps} />;
    }
  };

  // Style classes dựa trên type của toast
  const getTypeStyles = (type: TToastTypeValue): string => {
    switch (type) {
      case TOAST_TYPE.success:
        return "bg-primary text-white";
      case TOAST_TYPE.error:
        return "bg-red-500 text-white";
      case TOAST_TYPE.warning:
        return "bg-yellow-200 text-text";
      case TOAST_TYPE.info:
        return "bg-blue-500 text-white";
    }
  };

  // Classes cho animation
  const getAnimationClasses = (toastId: string): string => {
    const isExiting = exitingToasts.includes(toastId);

    const baseClasses = [
      // Base transition settings
      "transform",
      `transition-all duration-${TOAST_DURATION} ease-in-out`,

      // Entry animation (slide from right)
      "animate-slide-in-right",

      // Container styles
      "rounded-lg p-4 shadow-lg",
      "min-w-[300px] max-w-[400px]",
    ];

    // Exit animation classes
    const animationClasses = isExiting
      ? "opacity-0 translate-x-full" // Fade out + slide right
      : "opacity-100 translate-x-0"; // Normal state

    return [...baseClasses, animationClasses].join(" ");
  };

  return (
    // Container cho tất cả toasts
    <div className="fixed top-4 right-4 z-[60] flex flex-col gap-2">
      {toasts.map(toast => (
        <div
          data-testid="toast-test"
          id={`toast-${toast.id}`}
          key={toast.id}
          className={`
            ${getAnimationClasses(toast.id)}
            ${getTypeStyles(toast.type)}
          `}
        >
          {/* Toast content */}
          <div className="flex items-center gap-3">
            <ToastIcon toast={toast} />

            {/* Message */}
            <div className="flex-1 text-sm break-words">{toast.message}</div>

            {/* Close button */}
            <button
              data-testid="close-button-test"
              onClick={() => handleClose(toast.id)}
              className="shrink-0 rounded-full p-1 transition-colors duration-200
                hover:bg-black/10 focus:bg-black/10 focus:outline-none"
            >
              <XCircle size={20} />
            </button>
          </div>
        </div>
      ))}
    </div>
  );
};

export default Toast;
export { TOAST_DURATION };
export type { IToastProps };
