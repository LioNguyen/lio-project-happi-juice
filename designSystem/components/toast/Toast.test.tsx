import { screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { describe, expect, it, vi } from "vitest";

import { IToastMessage, TOAST_TYPE } from "@/domains/global";
import { renderWithTheme } from "@/shared/utils";
import Toast, { IToastProps, TOAST_DURATION } from "./Toast";

const mockCloseToast = vi.fn();

const defaultProps: IToastProps = {
  toasts: [],
  closeToast: mockCloseToast,
};

// Helper function to render Toast with mocked data
const renderToast = (toasts: Partial<IToastMessage>[]) => {
  const toastProps = toasts.map((toast, index) => ({
    id: `toast-${index}`,
    type: TOAST_TYPE.success,
    message: "Default message",
    showIcon: true,
    ...toast,
  })) as IToastMessage[];

  return renderWithTheme(<Toast {...defaultProps} toasts={toastProps} />);
};

// Helper to get toast element by message
const getToastByMessage = (message: string) => {
  return screen.getByText(message).closest('[data-testid="toast-test"]');
};

describe("Toast Component", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe("Basic Rendering", () => {
    // Snapshot Tests
    it("should match snapshot", () => {
      const { container } = renderToast([]);
      expect(container).toMatchSnapshot();
    });

    it("should not render any toasts when empty", () => {
      renderToast([]);

      expect(screen.queryByTestId("toast-test")).not.toBeInTheDocument();
    });

    it("should render toast with empty message", () => {
      renderToast([{ message: "" }]);

      const toast = screen.getByTestId("toast-test");
      expect(toast).toBeInTheDocument();
    });

    it("should render toast correctly", () => {
      const toastId = "test-id";
      renderToast([{ id: toastId, message: "First toast" }]);

      expect(screen.getByText("First toast")).toBeInTheDocument();
      expect(screen.getByText("First toast").closest(`#toast-${toastId}`)).toBeInTheDocument();
    });

    it("should render multiple toasts", () => {
      renderToast([{ message: "First toast" }, { message: "Second toast" }]);

      expect(screen.getByText("First toast")).toBeInTheDocument();
      expect(screen.getByText("Second toast")).toBeInTheDocument();
    });
  });

  describe("Toast Types and Styling", () => {
    it.each([
      [TOAST_TYPE.success, "bg-primary"],
      [TOAST_TYPE.error, "bg-red-500"],
      [TOAST_TYPE.warning, "bg-yellow-200"],
      [TOAST_TYPE.info, "bg-blue-500"],
    ])("should render %s toast with correct styling", (type, expectedClass) => {
      renderToast([{ type, message: `${type} toast` }]);

      const toast = getToastByMessage(`${type} toast`);
      expect(toast).toHaveClass(expectedClass);
    });
  });

  describe("Icon Display", () => {
    it.each([
      [TOAST_TYPE.success, "CheckCircle"],
      [TOAST_TYPE.error, "XCircle"],
      [TOAST_TYPE.warning, "AlertTriangle"],
      [TOAST_TYPE.info, "Info"],
    ])("should render correct icon for %s toast", (type, iconName) => {
      renderToast([{ type, message: "Test toast" }]);

      const icon = document.querySelector(`[data-testid="${iconName}-icon"]`);
      expect(icon).toBeInTheDocument();
    });

    it("should not render icon when showIcon is false", () => {
      renderToast([
        {
          type: TOAST_TYPE.success,
          message: "No icon toast",
          showIcon: false,
        },
      ]);

      const toast = getToastByMessage("No icon toast");
      expect(toast?.querySelector('[data-testid$="-icon"]')).not.toBeInTheDocument();
    });
  });

  describe("Close Button", () => {
    it("should render close button for each toast", () => {
      renderToast([{ message: "First toast" }, { message: "Second toast" }]);

      const closeButtons = screen.getAllByTestId("close-button-test");
      expect(closeButtons).toHaveLength(2);
    });

    it("should handle close button click", async () => {
      const toastId = "test-id";
      renderToast([{ id: toastId, message: "Closing toast" }]);

      const closeButton = screen.getByTestId("close-button-test");
      await userEvent.click(closeButton);

      await waitFor(
        () => {
          expect(mockCloseToast).toHaveBeenCalledWith(toastId);
        },
        { timeout: TOAST_DURATION + 100 },
      );
    });
  });

  describe("Animation Behavior", () => {
    it("should have entry animation classes by default", () => {
      renderToast([{ message: "Animated toast" }]);

      const toast = getToastByMessage("Animated toast");
      expect(toast).toHaveClass("animate-slide-in-right");
    });

    it("should apply exit animation classes when closing", async () => {
      renderToast([{ message: "Closing toast" }]);

      const closeButton = screen.getByTestId("close-button-test");
      await userEvent.click(closeButton);

      const toast = getToastByMessage("Closing toast");
      expect(toast).toHaveClass("opacity-0", "translate-x-full");
    });

    it("should remove toast after animation completes", async () => {
      renderToast([{ message: "Disappearing toast" }]);

      const closeButton = screen.getByTestId("close-button-test");
      await userEvent.click(closeButton);

      await waitFor(
        () => {
          expect(mockCloseToast).toHaveBeenCalled();
        },
        { timeout: TOAST_DURATION + 100 },
      );
    });
  });
});
