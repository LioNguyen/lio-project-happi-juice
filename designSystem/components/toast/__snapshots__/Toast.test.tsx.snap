// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Toast Component > Basic Rendering > should match snapshot 1`] = `
<div>
  <div
    class="fixed top-4 right-4 z-[60] flex flex-col gap-2"
  />
</div>
`;
